enhancements/api-review/add-new-CRD-ImageContentPolicy(ICP)-to-config.openshift.io.md

```diff
@@ -0,0 +1,260 @@
+[[registry]]
+  location = "registry.access.redhat.com/ubi8/"
+  insecure = false
+  blocked = false
+  # No mirror-by-digest-only = true configured
+  prefix = ""
+
+  [[registry.mirror]]
+    location = "example.io/example/ubi8-minimal"
+    insecure = false
+```
+
+### User Stories
+
+#### As a user, I would like to pull images from mirror using tag, without digests reference
+
+The user need to define multiple ImageContentPolicies, used by apps/manifests that don't use digests to pull
+the images when working with disconnected environments or pulling from registries that act as transparent pull-through proxy cache.
+
+The user can pull image without digest by configuring `allowMirrorByTags: true` in the ImageContentPolicy file.
+And create the ImageContentPolicy project. Once this is done, the images can be pulled from the mirrors without the digest referenced.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Implementing this enhancement requires changes in:
+- [openshift/api/config/v1](https://github.com/openshift/api/tree/master/config/v1): definition the schema and API of the ImageContentPolicy CRD.
+- [openshift/client-go](https://github.com/openshift/client-go), [openshift/cluster-config-operator](https://github.com/openshift/cluster-config-operator/pull/220): rebase the openshift/api version in these repositories to apply the new CRD
+to the cluster.
+- [openshift/runtime-utils/pkg/registries](https://github.com/openshift/runtime-utils/tree/master/pkg/registries): helper functions to edit `mirror-by-tags-only` in registries.conf.
+- [openshift/machine-config-operator](https://github.com/openshift/machine-config-operator): MCO needs watch for the ImageContentPolicy.
+- This [document](https://docs.google.com/document/d/11FJPpIYAQLj5EcYiJtbi_bNkAcJa2hCLV63WvoDsrcQ/edit?usp=sharing) keeps a list of components that use operator.openshift.io/v1alpha1 ImageContentSourcePolicy. Need to change those repositories to upgrade to ImageContentPolicy.
+
+#### Notes
+
+1. During the upgrade path, MCO can watch for both ImageContentPolicy and ImageContentSourcePolicy and create objects.
+2. The rejection of conflicting `allowMirrorByTags` values:<br/>
+   The validation requires the `allowMirrorByTags` to be same for the same `source`. MCO will fail on the conflicting values.
+3. The merge order of mirrors for the same source:<br/>
+   Order is preserving the relative order of the mirrors, if possible.
+   For example, 
+    - if policy A has mirrors `a, b, c` and policy B has mirrors `c, d, e`, the
+   mirrors will be used in the order `a, b, c, d, e`.
+   	- if policy A and policy B contain the mirrors cannot be merged by the relative order,
+	   they will be sorted in alphabetical order. If policy A has mirrors `a, d, e` and
+	   policy B has mirrors `b, c, e`, the `a, d` and `b, c` will be sorted in alphabetical order,
+	   mirrors will be used in the order `a, b, c, d, e`.
```

Comment:
@mtrmac Could you take a look, if this is the alphabetical order we expect? It is different from the current behavior.  Given the `a, d, e` and `b,c,e`, the current merge will return `a, b, d, c, e`

Reply:
> Could you take a look, if this is the alphabetical order we expect? It is different from the current behavior. Given the `a, d, e` and `b,c,e`, the current merge will return `a, b, d, c, e`.

It turns out I misremembered about how the code resolves ties. I’m sorry.

So this
> If we have to specify this, probably `s/cannot be merged/don’t specify or imply an order between a pair of mirrors/`.

is not correct.

---

Details:

Alphabetical order is only used to break loops, unordered _initial_ nodes (e.g. `a, c` vs. `b, c`), and order of edges going from a single node (`a, b` vs. `a, c`). Other unordered cases (e.g. the `c`, `d` node in the example above) are emitted based on their position in the queue of eligible nodes, i.e. on 1) the order their last predecessor was processed, and 2) in alphabetical order _within nodes_ added by the last predecessor.

I.e. in this case:
- (Ordered input is `a, b, c, d, e`. That doesn’t really matter.)
- Initial `noIncoming` list is `a, b`, items with no predecessor (in alphabetical order)
- `a` is output; that makes `d` eligible, and added to `noIncoming`, which is now `b, d`.
- `b` is output; that makes `c` eligible, and added to `noIncoming`, which is now `d, c` (Note: not `c, d`, because `noIncoming` is not re-sorted.)
- `d` is output, `e` is not eligible yet, noIncoming is now `c`
- `c` is output; that makes `e` eligible and added to `noIncoming`, which is now `e`
- `e` is output

We could fix this to resolve all ties alphabetically by sorting `noIncoming` on every iteration, but that would make this (without using more complex ordered set data structures) an `O(N^2 log N)` algorithm instead of an `O(N log N)` one. That might well be OK — who has even 10 mirrors for a single source? — but I don’t think it would make the ordering significantly more useful to users.

---

Adding
```go
		{
			name:   "Partial order",
			edges:  []string{"AD", "DE", "BC", "CE"},
			result: "ABDCE",
		},
```
to `openshift/runtime-utils/pkg/registries/topo_test.go:TestTopoGraph` verifies that the output is consistent WRT the order the 4 edges are specified (although it only tests 4 rotations, not all possible permutations); that is expected,  because both the original node list and the list of eligible outgoing edges of each node are sorted before being used.